pkg/planner/core/expression_rewriter.go

```diff
@@ -2780,11 +2797,16 @@ func decodeIndexKey(key []byte, tableID int64, tbl table.Table, loc *time.Locati
 		ret["index_id"] = indexID
 		idxValMap := make(map[string]any, len(targetIndex.Columns))
 		for i := 0; i < len(targetIndex.Columns); i++ {
-			dtStr, err := datumToJSONObject(&ds[i])
-			if err != nil {
-				return "", errors.Trace(err)
+			if types.NeedRestoredData(tps[i]) {
+				// Using sort key, better to show the key as HEX
+				idxValMap[targetIndex.Columns[i].Name.L] = fmt.Sprintf("%x", ds[i].GetString())
```

Comment:
This results in this:
```
mysql-8.0.11-TiDB-v8.2.0-alpha-571-g8de5fb9cf9> SELECT TIDB_DECODE_KEY('74800000000000008E5F72038000000000000001014100000000000000F8');
+---------------------------------------------------------------------------------+
| TIDB_DECODE_KEY('74800000000000008E5F72038000000000000001014100000000000000F8') |
+---------------------------------------------------------------------------------+
| {"handle":{"id":"1","name":"41"},"table_id":142}                                |
+---------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

The issue with this is that it is unclear that this is a hex literal instead of a actual string. Looking for a record with `id=1` and `name=41` yields zero results, while looking for `id=1` and `name=x041` does yield results.

```
mysql-8.0.11-TiDB-v8.2.0-alpha-571-g8de5fb9cf9> SELECT * FROM t9 WHERE id=1 AND name='41';
Empty set (0.00 sec)

mysql-8.0.11-TiDB-v8.2.0-alpha-571-g8de5fb9cf9> SELECT * FROM t9 WHERE id=1 AND name=0x41;
+----+------+------+
| id | name | c3   |
+----+------+------+
|  1 | A    |    1 |
+----+------+------+
1 row in set (0.00 sec)
```

I think this should be prefixed with `0x`.

```suggestion
				idxValMap[targetIndex.Columns[i].Name.L] = fmt.Sprintf("0x%x", ds[i].GetString())
```

It would be great if we could change `"id": "1"` to `"id": 1` and `"name": "0x41"` to `"name": 0x41`, but that's likely out of scope for this PR (although something similar is done for `table_id` within this PR). Also, I don't think JSON supports hex literals like this.